File: backend/schemas/analytics.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


OperatorLiteral = Literal[
    "eq",
    "neq",
    "contains",
    "starts_with",
    "in",
    "gt",
    "gte",
    "lt",
    "lte",
    "between",
    "on",
    "before",
    "after",
    "relative_range",
    "is_true",
    "is_false",
    "is_empty",
    "is_not_empty",
]

AggregateLiteral = Literal["sum", "avg", "min", "max", "count", "distinct_count"]
ModeLiteral = Literal["pivot", "table"]


class AnalyticsFieldMeta(BaseModel):
    key: str
    label: str
    kind: Literal["dimension", "measure"]
    data_type: Literal["string", "number", "boolean", "date", "datetime"]
    group: str
    description: str | None = None
    operators: list[OperatorLiteral] = Field(default_factory=list)
    allowed_aggregates: list[AggregateLiteral] = Field(default_factory=list)
    default_aggregate: AggregateLiteral | None = None
    is_linked_measure: bool = False


class AnalyticsStarterView(BaseModel):
    key: str
    label: str
    description: str
    subject_key: str
    config: dict[str, Any]


class AnalyticsSubjectMeta(BaseModel):
    key: str
    label: str
    description: str
    grain_label: str
    fields: list[AnalyticsFieldMeta] = Field(default_factory=list)
    default_table_fields: list[str] = Field(default_factory=list)
    default_values: list[dict[str, Any]] = Field(default_factory=list)


class AnalyticsCatalogResponse(BaseModel):
    subjects: list[AnalyticsSubjectMeta] = Field(default_factory=list)
    starter_views: list[AnalyticsStarterView] = Field(default_factory=list)
    executive_packs: list["AnalyticsExecutivePack"] = Field(default_factory=list)
    executive_filter_options: "AnalyticsExecutiveFilterOptions" = Field(default_factory=lambda: AnalyticsExecutiveFilterOptions())


class AnalyticsFilter(BaseModel):
    field: str
    op: OperatorLiteral
    value: Any = None
    value_to: Any = None


class AnalyticsSort(BaseModel):
    field: str
    direction: Literal["asc", "desc"] = "asc"


class AnalyticsValueSpec(BaseModel):
    key: str
    aggregate: AggregateLiteral | None = None
    alias: str | None = None


class AnalyticsQueryOptions(BaseModel):
    show_subtotals: bool = True
    show_grand_totals: bool = True
    hide_empty: bool = False
    hide_zero: bool = False
    row_limit: int = Field(default=200, ge=1, le=50000)
    column_limit: int = Field(default=50, ge=1, le=200)


class AnalyticsQueryRequest(BaseModel):
    subject_key: str
    mode: ModeLiteral = "pivot"
    rows: list[str] = Field(default_factory=list)
    columns: list[str] = Field(default_factory=list)
    values: list[AnalyticsValueSpec] = Field(default_factory=list)
    selected_fields: list[str] = Field(default_factory=list)
    filters: list[AnalyticsFilter] = Field(default_factory=list)
    sorts: list[AnalyticsSort] = Field(default_factory=list)
    options: AnalyticsQueryOptions = Field(default_factory=AnalyticsQueryOptions)


class AnalyticsExecutiveFilterBinding(BaseModel):
    fund_field: str | None = None
    date_field: str | None = None


ExecutiveVisualLiteral = Literal[
    "kpi",
    "line",
    "stacked_area",
    "bar",
    "grouped_bar",
    "stacked_bar",
    "donut",
    "ranked_bar",
    "table",
]


class AnalyticsExecutiveCard(BaseModel):
    key: str
    title: str
    description: str
    subject_key: str
    visual_type: ExecutiveVisualLiteral
    height: Literal["sm", "md", "lg"] = "md"
    query: AnalyticsQueryRequest
    filter_binding: AnalyticsExecutiveFilterBinding = Field(default_factory=AnalyticsExecutiveFilterBinding)
    direct_analysis_label: str = "직접 분석으로 열기"


class AnalyticsExecutiveSection(BaseModel):
    key: str
    label: str
    layout: Literal["kpi", "grid"] = "grid"
    cards: list[AnalyticsExecutiveCard] = Field(default_factory=list)


class AnalyticsExecutivePack(BaseModel):
    key: str
    label: str
    description: str
    sections: list[AnalyticsExecutiveSection] = Field(default_factory=list)


class AnalyticsOptionItem(BaseModel):
    value: str
    label: str


class AnalyticsExecutiveFilterOptions(BaseModel):
    funds: list[AnalyticsOptionItem] = Field(default_factory=list)


class AnalyticsResultField(BaseModel):
    key: str
    label: str
    kind: Literal["dimension", "measure"]
    data_type: Literal["string", "number", "boolean", "date", "datetime"]


class AnalyticsQueryMeta(BaseModel):
    subject_key: str
    subject_label: str
    mode: ModeLiteral
    grain_label: str
    execution_ms: int = 0
    truncated: bool = False
    warnings: list[str] = Field(default_factory=list)
    result_count: int = 0


class AnalyticsQueryResponse(BaseModel):
    meta: AnalyticsQueryMeta
    row_fields: list[AnalyticsResultField] = Field(default_factory=list)
    column_fields: list[AnalyticsResultField] = Field(default_factory=list)
    value_fields: list[AnalyticsResultField] = Field(default_factory=list)
    table_fields: list[AnalyticsResultField] = Field(default_factory=list)
    rows: list[dict[str, Any]] = Field(default_factory=list)
    grand_totals: dict[str, Any] = Field(default_factory=dict)


class AnalyticsSavedViewCreate(BaseModel):
    name: str
    description: str | None = None
    subject_key: str
    config: dict[str, Any]
    is_favorite: bool = False


class AnalyticsSavedViewUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    subject_key: str | None = None
    config: dict[str, Any] | None = None
    is_favorite: bool | None = None


class AnalyticsSavedViewResponse(BaseModel):
    id: int
    owner_user_id: int
    name: str
    description: str | None = None
    subject_key: str
    config: dict[str, Any]
    is_favorite: bool = False
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AnalyticsExportRequest(BaseModel):
    view_id: int | None = None
    query: AnalyticsQueryRequest | None = None
    file_name: str | None = None


class AnalyticsBatchQueryItem(BaseModel):
    key: str
    query: AnalyticsQueryRequest


class AnalyticsBatchQueryRequest(BaseModel):
    items: list[AnalyticsBatchQueryItem] = Field(default_factory=list)


class AnalyticsBatchQueryResult(BaseModel):
    key: str
    response: AnalyticsQueryResponse | None = None
    error: str | None = None


class AnalyticsBatchQueryResponse(BaseModel):
    results: list[AnalyticsBatchQueryResult] = Field(default_factory=list)

File: frontend/src/lib/api/analytics.ts
import { api } from './client'

export type AnalyticsFieldKind = 'dimension' | 'measure'
export type AnalyticsDataType = 'string' | 'number' | 'boolean' | 'date' | 'datetime'
export type AnalyticsMode = 'pivot' | 'table'
export type AnalyticsAggregate = 'sum' | 'avg' | 'min' | 'max' | 'count' | 'distinct_count'

export interface AnalyticsFieldMeta {
  key: string
  label: string
  kind: AnalyticsFieldKind
  data_type: AnalyticsDataType
  group: string
  description?: string | null
  operators: string[]
  allowed_aggregates: AnalyticsAggregate[]
  default_aggregate?: AnalyticsAggregate | null
  is_linked_measure: boolean
}

export interface AnalyticsSubjectMeta {
  key: string
  label: string
  description: string
  grain_label: string
  fields: AnalyticsFieldMeta[]
  default_table_fields: string[]
  default_values: AnalyticsValueSpec[]
}

export interface AnalyticsStarterView {
  key: string
  label: string
  description: string
  subject_key: string
  config: AnalyticsQueryRequest
}

export interface AnalyticsExecutiveFilterBinding {
  fund_field?: string | null
  date_field?: string | null
}

export type AnalyticsExecutiveVisualType =
  | 'kpi'
  | 'line'
  | 'stacked_area'
  | 'bar'
  | 'grouped_bar'
  | 'stacked_bar'
  | 'donut'
  | 'ranked_bar'
  | 'table'

export interface AnalyticsExecutiveCard {
  key: string
  title: string
  description: string
  subject_key: string
  visual_type: AnalyticsExecutiveVisualType
  height: 'sm' | 'md' | 'lg'
  query: AnalyticsQueryRequest
  filter_binding: AnalyticsExecutiveFilterBinding
  direct_analysis_label: string
}

export interface AnalyticsExecutiveSection {
  key: string
  label: string
  layout: 'kpi' | 'grid'
  cards: AnalyticsExecutiveCard[]
}

export interface AnalyticsExecutivePack {
  key: string
  label: string
  description: string
  sections: AnalyticsExecutiveSection[]
}

export interface AnalyticsOptionItem {
  value: string
  label: string
}

export interface AnalyticsExecutiveFilterOptions {
  funds: AnalyticsOptionItem[]
}

export interface AnalyticsCatalogResponse {
  subjects: AnalyticsSubjectMeta[]
  starter_views: AnalyticsStarterView[]
  executive_packs: AnalyticsExecutivePack[]
  executive_filter_options: AnalyticsExecutiveFilterOptions
}

export interface AnalyticsFilter {
  field: string
  op: string
  value?: unknown
  value_to?: unknown
}

export interface AnalyticsSort {
  field: string
  direction: 'asc' | 'desc'
}

export interface AnalyticsValueSpec {
  key: string
  aggregate?: AnalyticsAggregate | null
  alias?: string | null
}

export interface AnalyticsQueryOptions {
  show_subtotals: boolean
  show_grand_totals: boolean
  hide_empty: boolean
  hide_zero: boolean
  row_limit: number
  column_limit: number
}

export interface AnalyticsQueryRequest {
  subject_key: string
  mode: AnalyticsMode
  rows: string[]
  columns: string[]
  values: AnalyticsValueSpec[]
  selected_fields: string[]
  filters: AnalyticsFilter[]
  sorts: AnalyticsSort[]
  options: AnalyticsQueryOptions
}

export interface AnalyticsResultField {
  key: string
  label: string
  kind: AnalyticsFieldKind
  data_type: AnalyticsDataType
}

export interface AnalyticsQueryMeta {
  subject_key: string
  subject_label: string
  mode: AnalyticsMode
  grain_label: string
  execution_ms: number
  truncated: boolean
  warnings: string[]
  result_count: number
}

export interface AnalyticsQueryResponse {
  meta: AnalyticsQueryMeta
  row_fields: AnalyticsResultField[]
  column_fields: AnalyticsResultField[]
  value_fields: AnalyticsResultField[]
  table_fields: AnalyticsResultField[]
  rows: Record<string, unknown>[]
  grand_totals: Record<string, unknown>
}

export interface AnalyticsSavedView {
  id: number
  owner_user_id: number
  name: string
  description?: string | null
  subject_key: string
  config: AnalyticsQueryRequest
  is_favorite: boolean
  created_at: string
  updated_at: string
}

export interface AnalyticsBatchQueryItem {
  key: string
  query: AnalyticsQueryRequest
}

export interface AnalyticsBatchQueryRequest {
  items: AnalyticsBatchQueryItem[]
}

export interface AnalyticsBatchQueryResult {
  key: string
  response?: AnalyticsQueryResponse | null
  error?: string | null
}

export interface AnalyticsBatchQueryResponse {
  results: AnalyticsBatchQueryResult[]
}

export interface AnalyticsSavedViewCreate {
  name: string
  description?: string | null
  subject_key: string
  config: AnalyticsQueryRequest
  is_favorite?: boolean
}

export interface AnalyticsSavedViewUpdate {
  name?: string
  description?: string | null
  subject_key?: string
  config?: AnalyticsQueryRequest
  is_favorite?: boolean
}

export interface AnalyticsExportRequest {
  view_id?: number | null
  query?: AnalyticsQueryRequest | null
  file_name?: string | null
}

export async function fetchAnalyticsCatalog() {
  const { data } = await api.get<AnalyticsCatalogResponse>('/analytics/catalog')
  return data
}

export async function runAnalyticsQuery(payload: AnalyticsQueryRequest) {
  const { data } = await api.post<AnalyticsQueryResponse>('/analytics/query', payload)
  return data
}

export async function runAnalyticsBatchQuery(payload: AnalyticsBatchQueryRequest) {
  const { data } = await api.post<AnalyticsBatchQueryResponse>('/analytics/query-batch', payload)
  return data
}

export async function fetchAnalyticsViews() {
  const { data } = await api.get<AnalyticsSavedView[]>('/analytics/views')
  return data
}

export async function fetchAnalyticsView(id: number) {
  const { data } = await api.get<AnalyticsSavedView>(`/analytics/views/${id}`)
  return data
}

export async function createAnalyticsView(payload: AnalyticsSavedViewCreate) {
  const { data } = await api.post<AnalyticsSavedView>('/analytics/views', payload)
  return data
}

export async function updateAnalyticsView(id: number, payload: AnalyticsSavedViewUpdate) {
  const { data } = await api.put<AnalyticsSavedView>(`/analytics/views/${id}`, payload)
  return data
}

export async function deleteAnalyticsView(id: number) {
  const { data } = await api.delete<{ ok: boolean }>(`/analytics/views/${id}`)
  return data
}

export async function exportAnalyticsXlsx(payload: AnalyticsExportRequest) {
  const response = await api.post<Blob>('/analytics/export/xlsx', payload, { responseType: 'blob' })
  return response.data
}
